### DISCUSSION TITLE
Emulate grouping behaviour

### DISCUSSION THREAD

**Discussion by crides** [Z3Prover/z3#5725] [Q&A]
So I want to have regexes in z3 and have groups, but from what I can see right now that's not an option in z3, and I want to try to emulate groups. So the current approach is for every group, create a pair of ints that represents the start and end matching positions of the group, and assert that what's before, inside and after the group matches the string sliced by the pair of ints. Example:

```py
    # bf_g1, g1, af_g1 are re expressions, and s is the string to be matched
    g1_s, g1_e = Ints("g1_s g1_e")
    solv = Solver()
    solv.add(And(g1_s >= 0, g1_e >= 0, g1_s < Length(s), g1_e < Length(s)))
    solv.add(InRe(SubString(s, 0, g1_s), bf_g1))
    solv.add(InRe(SubString(s, g1_s, g1_e - g1_s), g1))
    solv.add(InRe(SubString(s, g1_e, Length(s) - g1_e), af_g1))
```

However, this only works standalone, but if there needs to be a pre-condition, then there's a problem. How can I turn this into an actual z3 expression so that it can be used in other expressions? Thanks!

**Comment by NikolajBjorner** [comment]
@veanes 

I am not sure what the issue of the question is.

Instead of using integers and substring that have to deal with cases where g1_e - g1_s is negative 
you can introduce auxiliary strings, s1, s2, s3 and assert s == Concat(s1, s2, s3) while matching s1 with bf_g1, etc.

InRe are z3 expressions so you can add them in nested formulas.

**Reply by crides** [reply]
Using substrings as variables can work, but my question is how would one get the values of `s1`, `s2` and `s3` and use them later, while being able to put the whole thing inside of some formula?

Example:
```py
# params
s = StringVal("abc")
bf_g1, g1, af_g1 = Re("a"), Re("b"), Re("c")

s1, s2, s3 = Strings("s1 s2 s3")
# get s1, s2, s3 such that...
And(s == Concat(s1, s2, s3), InRe(s1, bf_g1), InRe(s2, g1), InRe(s3, af_g1))
# and then...
prove(And(s1 == "a", s2 == "b", s3 == "c"))
```

Using a `Solver` is problematic because it's not guaranteed that the full regex (i.e. `Concat(bf_g1, g1, af_g1)`) matches `s`

**Reply by veanes** [reply]
There is a general law over all regexes `R` and `S` (because there are no _anchors_ or _lookarounds_ allowed in regexes here)
that, for all strings `u` and `v` : (essentially `L(Concat(R,S)) = Concat(L(R),L(S))`):

`InRe(u, R) and InRe(v, S) implies InRe(Concat(u,v), Concat(R,S))`

The other direction is of course not true if the suffix of R and prefix of S overlap. 
But the scenario above does not seem to need that. But I could also be misunderstanding something here. 

Adding support of **capture groups** to regexes is a rather tricky issue in general.
and the semantics is also unclear although practically useful
(e.g. what if a group occurs inside a loop etc... it is a bit of a pandora's box)
